mvpa/cosmo_pca.m

```diff
@@ -0,0 +1,102 @@
+function [ds, pca_params]=cosmo_pca(ds, varargin)
+% normalize dataset either by estimating or applying estimated parameters
+%
+% [ds, pca_params]=cosmo_pca(ds[, pca_params, pca_explained_count, 
+%                            pca_explained_ratio])
+%
+% Inputs
+%   ds            a dataset struct with field .samples of size PxQ, or a
+%                 numeric array of that size
+%   pca_params    previously estimated pca parameters using the 
+%                 'pca_params' output result from a previous call to this 
+%                 function
+%   pca_explained_count    retain only the first 'pca_explained_count' 
+%                          components
+%   pca_explained_ratio    retain the first components that explain 
+%                 'pca_explained_ratio' percent of the variance (value
+%                 between 0 and 1, where 1 retains all components)
+%
+% Output
+%   ds            a dataset struct similar to ds, but with .samples data
+%                 transformed using pca.
+%   params        estimated parameters for pca. These can be re-used for a 
+%                 second pca step of an independent dataset. For example, 
+%                 parameters can be estimated from a training dataset and 
+%                 then applied to a testing dataset
+%
+% #   For CoSMoMVPA's copyright information and license terms,   #
+% #   see the COPYING file distributed with CoSMoMVPA.           #
+
+opt=cosmo_structjoin(varargin);
+
+apply_params = isfield(opt,'pca_params');
+pca_explained_count = isfield(opt,'pca_explained_count');
+pca_explained_ratio = isfield(opt,'pca_explained_ratio');
+    
+%they are mutually exclusive
+if sum([apply_params,pca_explained_count,pca_explained_ratio])>1
+    error(['apply_params, pca_explained_count, pca_explained_ratio '...
+      'are mutually exclusive.']);
+end
+
+is_ds=isstruct(ds) && isfield(ds,'samples');
+
+if is_ds
+    samples=ds.samples;
+else
+    samples=ds;
+end
+
+if apply_params
+    pca_params=opt.pca_params;
+    coeff=pca_params.coeff;
+    mu=pca_params.mu;
+    retain=pca_params.retain;
+    %de-mean and multiply with previously computed coefficients
+    samples=bsxfun(@minus,samples,mu)*coeff;
```

Comment:
Re a check, when the input is a dataset: how about storing ds.a and ds.fa in pca_params, and when applying them, verify that they match? That would add another safety feature which may avoid applying pca to different data than it was estimated on.